editor/gui/editor_quick_open_dialog.cpp

```diff
@@ -45,6 +46,49 @@
 #include "scene/gui/texture_rect.h"
 #include "scene/gui/tree.h"
 
+Rect2i HighlightedLabel::get_substr_rect(const Vector2i &p_substr) {
+	Ref<Font> font = get_theme_font(SceneStringName(font));
+	int font_size = get_theme_font_size(SceneStringName(font_size));
+	Vector2i prefix = font->get_string_size(get_text().substr(0, p_substr.x), HORIZONTAL_ALIGNMENT_LEFT, -1, font_size);
+	prefix.y = 0;
+	Vector2i size = font->get_string_size(get_text().substr(p_substr.x, p_substr.y), HORIZONTAL_ALIGNMENT_LEFT, -1, font_size);
+	return { prefix, size };
+}
+
+void HighlightedLabel::add_highlight(const Vector2i &p_interval) {
+	if (p_interval.x != -1) {
+		highlights.append(get_substr_rect(p_interval));
+		queue_redraw();
+	}
+}
+
+void HighlightedLabel::reset_highlights() {
+	highlights.clear();
+	queue_redraw();
+}
+
+void HighlightedLabel::_notification(int p_notification) {
+	if (p_notification == NOTIFICATION_DRAW) {
+		if (highlights.is_empty()) {
+			return;
+		}
+
+		Vector2i offset = get_character_bounds(0).position;
```

Comment:
This also won't work, character `0` is not necessarily first on the left. You can copy code for `Label` draw to get initial offset instead.

Reply:
Just to make sure I'm not totally missing something, you're talking about this code https://github.com/godotengine/godot/blob/master/scene/gui/label.cpp#L475-L534? I'm not super happy with my current approach for extracting the necessary part of it which you can imagine the implementations for:
```c++
struct DrawParameters {
	Vector2 offset;
	int first_line = 0;
	int last_line = 0;
	int line_spacing = 0;
};

RID get_line_rid(int p_line) const;
Rect2 get_line_rect(int p_line) const;
Label::DrawParameters get_draw_parameters() const
```
would it be preferable to just recompute the total height, line range, vsep? or did you have something else entirely in mind?

Reply:
I never liked this approach of drawing boxes on labels. I did it this way originally because I was trying to touch the quick open dialog code as little as possible.

But seeing as big changes are being made to the dialog. Maybe an idea is drop this approach and instead only use richtext labels in the dialog. That way you just have to build a string instead of calculating finicky offsets and such. The performance of rich text is pretty good now too.

Reply:
Previously I thought that bbcode wouldn't be suitable but yeah turns out I missed a few options on RichTextLabel. I'm just going to change this to use bbcode and avoid this.

Reply:
OK well it's in the latest commit. I don't know if the label refactor is totally appropriate for this pr, although it does seem like an overall useful refactor now that I have a better understanding of the code there. I haven't tested it in any multi-line or rtl applications yet though.

Reply:
I was more thinking to use just the font color like visual studio code:
![image](https://github.com/user-attachments/assets/4d059b18-34e6-465b-ba68-cf73197296b0)

Reply:
Yeah, I agree that's the trade-off. I have a partially complete bbcode version of this branch which I can finish to compare but I'm also curious what @bruvzg thinks of the label changes. The bbcode approach would definitely be fewer loc changed.

Reply:
I have the branch ready here although I'm sorta kicking myself for not structuring these git commits better so I may touch them up to make the code compare easier between them https://github.com/a-johnston/godot/tree/fuzzy-search-rtl

I definitely like using the font color more than the bg color with richtextlabel; it ends up looking like:

<img width="559" alt="Screenshot 2024-10-27 at 5 39 50 PM" src="https://github.com/user-attachments/assets/baa4fad5-1595-47c6-beb2-ab9ab1bab0cc">

<img width="548" alt="Screenshot 2024-10-27 at 5 39 55 PM" src="https://github.com/user-attachments/assets/75f8e46c-2e34-4557-baf8-b806ebda4df7">

It still is a bit janky because the richtextlabel always wants to be multiline with no overrun handling, has an opaque default stylebox, and at least on my macbook the letters slightly jitter as the color tags maybe are affecting the positioning (I don't see this on my windows machine though). There's probably a better way to handle the stylebox stuff. Overall I personally think either is fine, and I didn't notice any performance difference between the two approaches.